flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironment.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.taskexecutor.partition;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.runtime.deployment.InputGateDeploymentDescriptor;
+import org.apache.flink.runtime.deployment.ResultPartitionDeploymentDescriptor;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.executiongraph.PartitionInfo;
+import org.apache.flink.runtime.io.network.api.writer.ResultPartitionWriter;
+import org.apache.flink.runtime.io.network.partition.PartitionProducerStateProvider;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.shuffle.ShuffleEnvironment;
+import org.apache.flink.runtime.taskexecutor.TaskExecutor;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * Job-aware variation of the {@link ShuffleEnvironment} interface.
+ *
+ * @see ShuffleEnvironment
+ */
+public interface JobAwareShuffleEnvironment<P extends ResultPartitionWriter, G extends InputGate> extends AutoCloseable {
+
+	/**
+	 * Start the internal related services upon {@link TaskExecutor}'s startup.
+	 *
+	 * @return a port to connect to the task executor for shuffle data exchange, -1 if only local connection is possible.
+	 */
+	int start() throws IOException;
+
+	/**
+	 * Factory method for the task's {@link ResultPartitionWriter}s.
+	 *
+	 * @parem jobId the job id, used for tracking which job the created partitions belongs to
+	 * @param taskName the task name, used for logs
+	 * @param executionAttemptID execution attempt id of the task
+	 * @param resultPartitionDeploymentDescriptors descriptors of the partition, produced by the task
+	 * @param outputGroup shuffle specific group for output metrics
+	 * @param buffersGroup shuffle specific group for buffer metrics
+	 * @return collection of the task's {@link ResultPartitionWriter}s
+	 */
+	Collection<P> createResultPartitionWriters(
+		JobID jobId,
```

Comment:
In #8485 , Andrey introduced shuffle context to cover `taskName`, `executionAttemptID` and `MetricGroup`. If this context could also cover `jobId` then we could reduce this method definition as now.

Reply:
yes, that's the main argument against doing so. It'S a lose-lose situation, as we either
a) introduce a jobID into the shuffle service API, which we don't really want
b) extend the ShuffleService interface, overload `createResultPartitionWriters` with a variant that accepts a JobID, and have the other method fail with an exception
c) duplicate most of the interface to prevent downside of b.